athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/common/PartitionUtil.java

```diff
@@ -0,0 +1,212 @@
+        validatePartitionColumnTypes(partitionColumns);
+        String partitionPattern = table.getParameters().get(PARTITION_PATTERN_KEY);
+        // Check to see if there is a partition pattern configured for the Table by the user
+        // if not, it returns empty value
+        if (partitionPattern == null || partitionPattern.isBlank()) {
+            return Optional.empty();
+        }
+        Matcher partitionMatcher = PARTITION_PATTERN.matcher(partitionPattern);
+        String folderMatchingRegexPattern = partitionPattern;
+        while (partitionMatcher.find()) {
+            folderMatchingRegexPattern = folderMatchingRegexPattern.replace("{" + partitionMatcher.group(1) + "}", VARCHAR_OR_STRING_REGEX);
+        }
+        return Optional.of(folderMatchingRegexPattern);
+    }
+
+    /**
+     * Return a list of storage partition(column name, column type and value)
+     *
+     * @param partitionPattern Name of the bucket
+     * @param folderModel      partition folder name
+     * @param folderNameRegEx  folder name regular expression
+     * @param partitionColumns partition column name list
+     * @return List of storage partition(column name, column type and value)
+     */
+    public static List<PartitionColumnData> getStoragePartitions(String partitionPattern, String folderModel, String folderNameRegEx, List<Column> partitionColumns)
+    {
+        List<PartitionColumnData> partitions = new ArrayList<>();
+        String[] partitionPatternParts = partitionPattern.split("/");
+        String[] regExParts = folderNameRegEx.split("/");
+        String[] folderParts = folderModel.split("/");
+        if (folderParts.length >= regExParts.length) {
+            for (int i = 0; i < regExParts.length; i++) {
+                Matcher matcher = Pattern.compile(regExParts[i]).matcher(folderParts[i]);
+                if (matcher.matches() && matcher.groupCount() > 0) {
+                    String partitionColumn;
+                    String columnValue;
+                    if (matcher.group(0).equals(matcher.group(1))) { // Non-hive partition (e.g, /{partitionKey1}//{partitionKey2})
+                        Matcher nonHivePartitionPatternMatcher = PARTITION_PATTERN.matcher(partitionPatternParts[i]);
+                        if (nonHivePartitionPatternMatcher.matches()) {
+                            partitionColumn = nonHivePartitionPatternMatcher.group(1);
+                        }
+                        else { // unknown partition layout
+                            throw new IllegalArgumentException("Unsupported partition layout pattern. partition.pattern is " + partitionPattern);
+                        }
+                    }
+                    else if (matcher.group(0).contains("=")) { // Hive partition (e.g. /folderName1={partitionKey1}/folderName2={partitionKey2})
+                        partitionColumn = matcher.group(0).substring(0, matcher.group(0).indexOf("="));
+                    }
```

Comment:
Why does the code care about what "kind" of partitioning it is? The whole point of using a pattern is so that the user can use any kind of partitioning scheme they want.